Parse ```self``` as ```$_:ident```

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->

I tried this code:

```rust
macro_rules! foo { ($_:ident) => {} }

pub fn foo() {
    foo!(self);
}
```

I expected to see this happen: no error

Instead, this happened:

```
test.rs:4:5: error: Failed to match any rule within macro
    1 | macro_rules! foo { ($_:ident) => {} }
      | ~~~~~~~~~~~
......
    4 |     foo!(self);
      |     ^~~
```

### Meta

git: ```bb1e40da864addfc2d8dfc310928b7a4e7b13185```

--- a/gcc/rust/expand/rust-macro-expand.cc
+++ b/gcc/rust/expand/rust-macro-expand.cc
@@ -378,7 +378,7 @@ MacroExpander::match_fragment (Parser<MacroInvocLexer> &parser,
       break;
 
     case AST::MacroFragSpec::IDENT:
-      parser.parse_identifier_pattern ();
+      parser.parse_identifier_or_keyword_token ();
       break;
 
     case AST::MacroFragSpec::LITERAL:

--- a/gcc/rust/lex/rust-token.cc
+++ b/gcc/rust/lex/rust-token.cc
@@ -57,6 +57,40 @@ token_id_to_str (TokenId id)
     }
 }
 
+/* checks if a token is a keyword */
+bool
+token_id_is_keyword (TokenId id)
+{
+  switch (id)
+    {
+#define RS_TOKEN_KEYWORD(name, _) case name:
+#define RS_TOKEN(a, b)
+      RS_TOKEN_LIST return true;
+#undef RS_TOKEN_KEYWORD
+#undef RS_TOKEN
+    default:
+      return false;
+    }
+}
+
+/* gets the string associated with a keyword */
+const char *
+token_id_keyword_string (TokenId id)
+{
+  switch (id)
+    {
+#define RS_TOKEN_KEYWORD(id, str)                                              \
+  case id:                                                                     \
+    return str;
+#define RS_TOKEN(a, b)
+      RS_TOKEN_LIST
+#undef RS_TOKEN_KEYWORD
+#undef RS_TOKEN
+    default:
+      return nullptr;
+    }
+}
+
 const char *
 get_type_hint_string (PrimitiveCoreType type)
 {

--- a/gcc/rust/lex/rust-token.h
+++ b/gcc/rust/lex/rust-token.h
@@ -226,6 +226,12 @@ get_token_description (TokenId id);
  * x-macros */
 const char *
 token_id_to_str (TokenId id);
+/* checks if a token is a keyword */
+bool
+token_id_is_keyword (TokenId id);
+/* gets the string associated with a keyword */
+const char *
+token_id_keyword_string (TokenId id);
 // Get type hint description as a string.
 const char *
 get_type_hint_string (PrimitiveCoreType type);

--- a/gcc/rust/parse/rust-parse-impl.h
+++ b/gcc/rust/parse/rust-parse-impl.h
@@ -999,6 +999,24 @@ Parser<ManagedTokenSource>::parse_delim_token_tree ()
     }
 }
 
+// Parses an identifier/keyword as a Token
+template <typename ManagedTokenSource>
+std::unique_ptr<AST::Token>
+Parser<ManagedTokenSource>::parse_identifier_or_keyword_token ()
+{
+  const_TokenPtr t = lexer.peek_token ();
+
+  if (t->get_id () == IDENTIFIER || token_id_is_keyword (t->get_id ()))
+    {
+      lexer.skip_token ();
+      return std::unique_ptr<AST::Token> (new AST::Token (std::move (t)));
+    }
+  else
+    {
+      return nullptr;
+    }
+}
+
 /* Parses a TokenTree syntactical production. This is either a delimited token
  * tree or a non-delimiter token. */
 template <typename ManagedTokenSource>

--- a/gcc/rust/parse/rust-parse.h
+++ b/gcc/rust/parse/rust-parse.h
@@ -148,6 +148,7 @@ public:
   std::vector<std::unique_ptr<AST::LifetimeParam> > parse_lifetime_params ();
   AST::Visibility parse_visibility ();
   std::unique_ptr<AST::IdentifierPattern> parse_identifier_pattern ();
+  std::unique_ptr<AST::Token> parse_identifier_or_keyword_token ();
   std::unique_ptr<AST::TokenTree> parse_token_tree ();
   std::tuple<AST::SimplePath, std::unique_ptr<AST::AttrInput>, Location>
   parse_attribute_body ();

--- a/gcc/testsuite/rust/compile/macro-issue2192.rs
+++ b/gcc/testsuite/rust/compile/macro-issue2192.rs
@@ -1,0 +1,7 @@
+macro_rules! foo {
+    ($a:ident) => {}
+}
+
+pub fn bar() {
+    foo!(self);
+}

